edb/server/protocol/auth/scram.py

```diff
@@ -0,0 +1,316 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2022-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import base64
+import collections
+import hashlib
+import http
+import os
+import time
+from jwcrypto import jwt
+
+from edgedb import scram
+
+from edb.common import debug
+from edb.common import markup
+
+
+SESSION_TIMEOUT = 30
+SESSION_HIGH_WATER_MARK = SESSION_TIMEOUT * 10
+sessions = collections.OrderedDict()
+
+
+def handle_request(request, response, server):
+    if not request.authorization:
+        response.status = http.HTTPStatus.UNAUTHORIZED
+        response.custom_headers["WWW-Authenticate"] = "SCRAM-SHA-256"
+        return
+
+    scheme, _, auth_str = request.authorization.decode("ascii").partition(" ")
+    if scheme != "SCRAM-SHA-256":
+        response.body = (
+            b"Client selected an invalid SASL authentication mechanism"
+        )
+        response.status = http.HTTPStatus.UNAUTHORIZED
+        response.custom_headers["WWW-Authenticate"] = "SCRAM-SHA-256"
+        return
+
+    data = None
+    sid = None
+    try:
+        for kv_str in auth_str.split():
+            key, _, value = kv_str.rstrip(",").partition("=")
+            if key == "data":
+                data = base64.b64decode(value.strip('"')).strip()
+            elif key == "sid":
+                sid = value.strip('"')
+        if data is None:
+            raise ValueError("Malformed SCRAM message: data is missing")
+    except Exception as ex:
+        if debug.flags.server:
+            markup.dump(ex)
+        response.body = str(ex).encode("ascii")
+        response.status = http.HTTPStatus.BAD_REQUEST
+        response.close_connection = True
+        return
+
+    if (
+        not server.get_jwe_key().has_public
+        or not server.get_jws_key().has_private
```

Comment:
Should `has_public` and `has_private` be the other way around here? In `server.py` the `jwe_key` has the private key, and `jws_key` has the public key: https://github.com/edgedb/edgedb/blob/master/edb/server/server.py#L1611-L1628

Reply:
Oh this was intentional - in `edb/server/server.py` it uses the keys for decryption and signature verification, `scram.py` instead uses them for signing and encryption. That is to say, the remote instance must be a "development" instance, which has both private and public keys in both JOSE keys. Production instances should not usually use the `--jose-key-mode=generate` and have their own ways to generate/utilize the JOSE keys, like the EdgeDB Cloud instances, and in this case, HTTP SCRAM login is disabled (I think that's okay?).